# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Hashim123132/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Lane 1 = Ranking Signals Analysis



I'm choosing Lane 1 (Ranking Signal Analysis) as my provisional lane because I want to understand which safe content and search signals are associated with search visibility and clicks. In the starter dataset, 54.2% of pages fall into the "down" trend bucket, suggesting there is meaningful variation to investigate. This makes Ranking Signal Analysis a good starting point for building evidence-based insights before developing more advanced models later.

In [20]:
import pandas as pd

df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")

print(df.trend_direction.value_counts().to_dict())

{'down': 16262, 'stable': 5962, 'up': 4388, 'new': 2236, 'flat': 1152}


## 2. The question: decision, action, cost of a wrong call

Decision improved: Identifying which search and content signals should be prioritized to help protect or improve search visibility and clicks.

Who acts: An SEO specialist or content editor. If a company does not have one, a web developer or website owner can use the report to decide what to review first.

Frame: This project will analyze anonymized search and engagement data to identify which signals are most strongly associated with search performance. The goal is to support better SEO decisions with evidence rather than assumptions. A wrong recommendation can waste time reviewing low-impact pages while missing pages that are gradually losing visibility and should have been prioritized.

In [21]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

The slice holds 30,000 pages across 32 clients. Of those, 54.2% (16,262) sit in the down bucket — movement is the common case, not a rare event. Median average position is ≈ 11.4 (about the page-one/two boundary) while median CTR is 0.07%, so position-tier and a volume floor are required before reading any signal.

In [22]:
import pandas as pd
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")

print(len(df), "pages,", df.client_id.nunique(), "clients")
print("54.2% in down bucket:", (df.trend_direction == "down").mean().round(3))
print("median position:", df.loc[df.avg_position > 0, "avg_position"].median())

30000 pages, 32 clients
54.2% in down bucket: 0.542
median position: 11.4


## 4. Careful words: what I can and can't claim

Can claim: observed/measured associations, directional findings ("tends to go with"), decision-support for editors.
Cannot claim: causal proof (no experiment here), "predicting Google's algorithm," that a refresh caused a recovery, reconstructing raw queries/URLs/clients, or semantic insight.

In [23]:
import pandas as pd
df = pd.read_csv("../../data/raw/content_refresh_anonymized.csv")

print(list(df.columns))

['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']


## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.